README.md
# Simple Zeppelin Scala

![1.0.0](https://img.shields.io/badge/version-1.0.0-brightgreen) ![date](https://img.shields.io/badge/release--date-17%20Jan%202022-brightgreen)

This is an OTL interpreter for Apache Zeppelin. This interpreter works in tandem with OT Simple Rest and Apache Spark backend. REST Server collects OTL queries and converts them into Spark SQL queries with full support of all features of [SuperDispatcher](https://github.com/ISGNeuro/dispatcher).
## Prerequisites

- Java 1.8.1
- Sbt 1.3.8
- Scala 2.11.12
- ot_simple_connector_scala 1.1.0

## Installation

0. Let the `$ZEPPELIN_HOME` is the installation path of Apache Zeppelin.
1. Create subfolder `otl` in `$ZEPPELIN_HOME/interpreter`:
```
cd $ZEPPELIN_HOME/interpreter
mkdir otl
```
2. Copy `interpreter-setting.json` from this project to `$ZEPPELIN_HOME/interpreter/otl`. Copy `zeppelin-site.xml` to `$ZEPPELIN_HOME/conf`. If files are already existed, be careful about overwriting your local settings.
```
cp interpreter-setting.json $ZEPPELIN_HOME/interpreter/spl
cp zeppelin-site.xml $ZEPPELIN_HOME/conf
```
3. Make assembly of the current project. It's strongly recommended to make assembly rather than package due to external dependencies. Then copy jar-file of assembly to `$ZEPPELIN_HOME/interpreter/otl` and restart zeppelin daemon.

4. After restart, open Zeppelin and add a new interpreter with name "otl".

## Configuration

1. You should specify correct path to OT Simple Rest in the following interpreter properties in Zeppelin:

- `OTP.rest.host` (ex.: 192.168.4.65)
- `OTP.rest.port` (ex.: 50000)
- `OTP.rest.auth.username` (ex.: admin)
- `OTP.rest.auth.password` (ex.: 12345678)
- `OTP.rest.cache.host` (ex.: 192.168.4.65, usually the same as `OTP.rest.host`)
- `OTP.rest.cache.port` (ex.: 80, usually 80)

2. Optional properties:

- `OTP.query.timeout`. Timeout (seconds) for waiting for the result. Default value is 60.
- `OTP.query.ttl`. "Time-to-live" of cache of your query in dispatcher base. Usually need when multiple users produce the same query simultaneously. Default value is 60.
- `OTP.query.convert_time` ("true" or "false"). If converting from seconds to milliseconds required for '_time' column. Older Apache Zeppelin versions require timestamp in milliseconds for the correct visualization. Default value: "true"

## Usage

#### Writing queries

Just open Zeppelin notebook and write OTL queries.

Query:
```
%otl
| makeresults count=5 | streamstats count as cnt | eval squared = cnt * cnt
```
Result:
```
+----------+---+-------+
|     _time|cnt|squared|
+----------+---+-------+
|1605871742|  1|      1|
|1605871742|  2|      4|
|1605871742|  3|      9|
|1605871742|  4|     16|
|1605871742|  5|     25|
+----------+---+-------+
```

#### Form templates

You can use Zeppelin templates in OTL query to create form inputs.

Query:
```
| __read__ path=omds_well_adku
| search wellNum=${Well}
| timechart span=${Span=1d,1h|2h|1d|1w} mean(adkuLiquidDebit) as debit
| fields - 0
```

![Form templates](docs/img-templates.png)

#### Tokens in query

You can use variables previously saved in resource pool within OTL query.

1. Put object to resource pool:

Scala cell:
```scala
%spark
z.put("rename", "newname")
```

2. Use object value as token in OTL query.

Query:
```
| __read__ path=omds_well_adku
| search wellNum=13 AND adkuLiquidDebit=*
| timechart span=1d mean(adkuLiquidDebit) as $rename$
| fields - 0
```
Result:
```
+----------+-------+
|     _time|newname|
+----------+-------+
|1572566400|  108.5|
|1572739200|  108.0|
|1573084800|  117.0|
|1573171200|  110.0|
| ........ | ..... |
|1574380800|  113.5|
|1574467200|  106.0|
|1574553600|   96.0|
+----------+-------+
```

## Tested environment
- Apache Zeppelin 0.8.2
- Apache Spark 2.4.3
- OT Simple REST 1.6.0
- OT Super Dispatcher 1.2.0
- OT Scalarest 0.1.0

## Authors

- Nikolay Ryabykh (nryabykh@isgneuro.com)
- Sergey Ermilov (sermilov@isgneuro.com)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.1] 2023-01-26
### Fixed
- no time to explain

## [1.2.0] 2022-2-9
### Added
- OTLInterpreterMock class for local testing
### Changed
- tests
- zeppelin-interpreter 0.10.0

## [1.1.0] 2022-2-3
### Changed
- project structure
- some tests
### Removed
- **earliest**, **latest** functions support


## [1.0.0] 2022-1-17
#Changed
- Initial fork and refactor of **zeppelin-spl** project: https://github.com/ISGNeuroTeam/zeppelin-spl
- Some regex changed
- changed exception handling
#Added
- this CHANGELOG


src/test/scala/com/isgneuro/zeppelinotl/OTLInterpreterMock.scala
package com.isgneuro.zeppelinotl

import com.isgneuro.otp.connector.Core.Dataset
import com.isgneuro.otp.connector.Core
import com.isgneuro.otp.connector.utils.Exceptions.QueryExecutionException
import com.isgneuro.zeppelinotl.OTLInterpreterMock.DatasetMock

import java.util.Properties
import scala.util.{ Failure, Success, Try }

class OTLInterpreterMock(properties: Properties) extends OTLInterpreter(properties) {

  override def createJobAndReturnDataset(query: String, connectionInfo: Core.ConnectionInfo, username: String, password: String): Try[Dataset] = {
    query match {
      case str if str.equals("| makeresults ") =>
        val datasetJsonLines = Seq("""{"_time":1643889830}""")
        val datasetSchema = """`_time` BIGINT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("makeresults count=2 | eval x=2 | eval y=x*x | fields x, y") =>
        val datasetJsonLines = Seq("""{"x":2,"y":4}""", """{"x":2,"y":4}""")
        val datasetSchema = """`x` INT,`y` INT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("makeresults count=2 | streamstats count as x | stats list(x) as xs") =>
        val datasetJsonLines = Seq("""{"xs":[1,2]}""")
        val datasetSchema = """`xs` ARRAY<BIGINT>"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("| makeresults count=3 | streamstats count as x | eval y=x*x | fields x, y") =>
        val datasetJsonLines = Seq("""{"x":1,"y":1}""", """{"x":2,"y":4}""", """{"x":3,"y":9}""")
        val datasetSchema = """`x` BIGINT,`y` BIGINT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("""
                               | makeresults count=1000000
                               | streamstats count as x
                               | eval y = x * 20
                               | eventstats min(y) as miny
                               | stats min(miny) as minminy""".replace("\n", "")) =>
        val datasetJsonLines = Seq("""{"minminy":20}""")
        val datasetSchema = """`minminy` BIGINT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("| makeresults count=3 | where fields + 0") =>
        Failure(new QueryExecutionException("""[SearchId:228488] Error in  'where' command"""))

      case str if str.equals("| makeresults  count =  2 | eval x=1 | fields x") =>
        val datasetJsonLines = Seq("""{"x":1}""", """{"x":1}""")
        val datasetSchema = """`x` INT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("| makeresults count=10 | streamstats count as x | eval y=1 | fields x, y") =>
        val datasetJsonLines = Seq("""{"x":1,"y":1}""", """{"x":2,"y":1}""", """{"x":3,"y":1}""", """{"x":4,"y":1}""",
          """{"x":5,"y":1}""", """{"x":6,"y":1}""", """{"x":7,"y":1}""", """{"x":8,"y":1}""", """{"x":9,"y":1}""",
          """{"x":10,"y":1}""")
        val datasetSchema = """`x` BIGINT,`y` INT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

      case str if str.equals("| makeresults count =3 | eval a = 1 | search a<0 ") =>
        val datasetJsonLines = Seq.empty[String]
        val datasetSchema = """`_time` BIGINT,`a` INT"""
        Success(DatasetMock(datasetSchema, datasetJsonLines))

    }
  }

}

object OTLInterpreterMock {

  case class DatasetMock(datasetSchema: String, datasetLines: Seq[String]) extends Dataset {
    override val schema: String = datasetSchema
    override val jsonLines: Seq[String] = datasetLines
  }

}

src/test/scala/com/isgneuro/zeppelinotl/OTLInterpreterTest.scala
package com.isgneuro.zeppelinotl

import java.util.Properties
import org.apache.zeppelin.interpreter.InterpreterContext
import org.apache.zeppelin.resource.LocalResourcePool
import org.scalatest.{ FlatSpec, Matchers }

class OTLInterpreterTest extends FlatSpec with Matchers {

  val propertiesMap: Map[String, String] = Map(
    "OTP.rest.host" -> "192.168.4.65",
    "OTP.rest.port" -> "80",
    "OTP.rest.auth.username" -> "admin",
    "OTP.rest.auth.password" -> "12345678",
    "OTP.rest.cache.host" -> "192.168.4.65",
    "OTP.rest.cache.port" -> "80",
    "OTP.query.timeout" -> "60",
    "OTP.query.ttl" -> "60",
    "OTP.query.convertTime" -> "false")

  val properties: Properties = {
    val properties = new Properties()
    propertiesMap.foreach { case (k, v) => properties.setProperty(k, v) }
    properties
  }

  val interpreterContext: InterpreterContext = new InterpreterContext.Builder().build

  "Mock" should "work fine" in {
    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults ", interpreterContext)

    val expected = "%table _time\n1643889830"

    result.toString should be(expected)

  }

  it should "parse json lines from dataset and convert them to appropriate format" in {
    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("makeresults count=2 | eval x=2 | eval y=x*x | fields x, y", interpreterContext)

    val expected = "%table x\ty\n2\t4\n2\t4"

    result.toString should be(expected)
  }

  it should "convert List to correct string without 'List' word in resulting dataset" in {
    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("makeresults count=2 | streamstats count as x | stats list(x) as xs", interpreterContext)

    val expected = "%table xs\n(1, 2)"

    result.toString should be(expected)
  }

  it should "get query as string, run Connector, get and parse resulting dataset" in {
    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults count=3 | streamstats count as x | eval y=x*x | fields x, y", interpreterContext)

    val expected = "%table x\ty\n1\t1\n2\t4\n3\t9"

    result.toString should be(expected)
  }

  it should "return an error if query fails due to timeout" in {

    properties.setProperty("OTP.query.timeout", "-1")

    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("""
                                             | makeresults count=1000000
                                             | streamstats count as x
                                             | eval y = x * 20
                                             | eventstats min(y) as miny
                                             | stats min(miny) as minminy
      """.replace("\n", ""), interpreterContext)
    properties.setProperty("OTP.query.timeout", "60")

    result.code().toString should be("ERROR")
  }

  it should "return an error if query fails due to execution error" in {

    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults count=3 | where fields + 0", interpreterContext)

    result.code().toString should be("ERROR")

  }

  it should "convert _time column to millisecs in resulting DF according to OTP.query.convertTime property" in {

    properties.setProperty("OTP.query.convertTime", "true")

    val result1 = new OTLInterpreterMock(properties).interpret("| makeresults ", interpreterContext)
    val expected1 = "%table _time\n1643889830000"

    properties.setProperty("OTP.query.convertTime", "false")

    val result2 = new OTLInterpreterMock(properties).interpret("| makeresults ", interpreterContext)
    val expected2 = "%table _time\n1643889830"

    result1.toString should be(expected1)
    result2.toString should be(expected2)
  }

  it should "get tokens' values from resource pool and put them into query" in {
    val query = "| makeresults  count = $count$ | eval x=1 | fields x"

    interpreterContext.setResourcePool(new LocalResourcePool("LocalResourcePool"))
    interpreterContext.getResourcePool.put("count", "2")
    //val ctxRP = InterpreterContextHelper.setResourcePool(interpreterContext)

    //ctxRP.getResourcePool.put("count", "2")

    val expected = "| makeresults  count = 2 | eval x=1 | fields x"

    Query(query).setTokens(interpreterContext.getResourcePool).query should be(expected)

    interpreterContext.getResourcePool.remove("count")
  }

  it should "execute corrected query with tokens' values from resource pool" in {

    interpreterContext.setResourcePool(new LocalResourcePool("LocalResourcePool"))
    interpreterContext.getResourcePool.put("count", "2")

    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults  count =  $count$ | eval x=1 | fields x", interpreterContext)

    val expected = "%table x\n1\n1"

    interpreterContext.getResourcePool.remove("count")

    result.toString should be(expected)

  }

  it should "limit resulting dataset to value of maxResultRows" in {

    properties.setProperty("OTP.query.maxResultRows", "3")

    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults count=10 | streamstats count as x | eval y=1 | fields x, y", interpreterContext)

    val expected = "%table x\ty\n1\t1\n2\t1\n3\t1"

    result.toString should be(expected)

    properties.remove("OTP.query.maxResultRows")

  }

  it should "return a message if resulting dataset is empty" in {
    val mockInterpreter = new OTLInterpreterMock(properties)

    val result = mockInterpreter.interpret("| makeresults count =3 | eval a = 1 | search a<0 ", interpreterContext)
    val expected = "%text Resulting dataset is empty"

    result.toString should be(expected)
  }
}


src/test/scala/com/isgneuro/zeppelinotl/InterpreterContextHelper.scala
package com.isgneuro.zeppelinotl

import org.apache.zeppelin.interpreter.InterpreterContext
import org.apache.zeppelin.resource.{ LocalResourcePool, ResourcePool }

object InterpreterContextHelper {
  /*def setResourcePool(ctx: InterpreterContext, rp: ResourcePool): InterpreterContext = {
    new InterpreterContext(
      ctx.getNoteId,
      ctx.getParagraphId,
      ctx.getReplName,
      ctx.getParagraphTitle,
      ctx.getParagraphText,
      ctx.getAuthenticationInfo,
      ctx.getConfig,
      ctx.getGui,
      ctx.getNoteGui,
      ctx.getAngularObjectRegistry,
      rp,
      ctx.getRunners,
      ctx.out)
  }

  def setResourcePool(ctx: InterpreterContext): InterpreterContext = {
    val lrp = new LocalResourcePool("LocalResourcePool")
    setResourcePool(ctx, lrp)
  }*/
}


src/main/scala/com/isgneuro/zeppelinotl/Query.scala
package com.isgneuro.zeppelinotl

import com.isgneuro.otp.connector.utils.Datetime.getCurrentTime
import org.apache.zeppelin.resource.ResourcePool

import java.text.SimpleDateFormat
import scala.util.{ Failure, Success, Try }

case class Query(query: String) {

  /**
   * Extract tokens from resource pool and put them into query
   *
   * @param pool - Resource pool from interpreter context. It should contain values for tokens used in query
   * @return - New query with tokens' values embedded
   */
  private[zeppelinotl] def setTokens(pool: ResourcePool): Query = {
    case class Token(full: String, name: String)

    val newQuery: String = """\$(.*?)\$""".r.findAllIn(query).matchData
      .map(x => Token(x.group(0), x.group(1)))
      .foldLeft(query) {
        case (acc, tok) =>
          val tokenValue = Try(pool.get(tok.name).get.toString) match {
            case Success(tv) => tv
            case Failure(ex) => throw new NoSuchElementException(s"Caught exception '${ex.getMessage}', Cannot get token ${tok.name} from resource pool")
          }
          acc.replaceAllLiterally(tok.full, tokenValue)
      }
    Query(newQuery)
  }
}



src/main/scala/com/isgneuro/zeppelinotl/OTLInterpreter.scala
package com.isgneuro.zeppelinotl

import com.isgneuro.otp.connector.Core.{ ConnectionInfo, Dataset }
import com.isgneuro.otp.connector.Core.{ ConnectionInfo, Dataset }
import com.isgneuro.otp.connector.utils.Datetime.{ convertTimeToMillis, getCurrentTime }
import com.isgneuro.otp.connector.Connector
import org.apache.zeppelin.interpreter.Interpreter.FormType
import org.apache.zeppelin.interpreter.InterpreterResult.{ Code, Type }
import org.apache.zeppelin.interpreter.{ Interpreter, InterpreterContext, InterpreterResult }
import org.json4s.DefaultFormats
import org.json4s.native.JsonMethods.parse

import java.util.Properties
import scala.util.{ Failure, Success, Try }

class OTLInterpreter(properties: Properties) extends Interpreter(properties) {
  val host: Option[String] = Option(properties.getProperty("OTP.rest.host"))
  val port: Option[String] = Option(properties.getProperty("OTP.rest.port"))
  val username: Option[String] = Option(properties.getProperty("OTP.rest.auth.username"))
  val password: Option[String] = Option(properties.getProperty("OTP.rest.auth.password"))
  val datasetHost: Option[String] = Option(properties.getProperty("OTP.rest.cache.host"))
  val datasetPort: String = Option(properties.getProperty("OTP.rest.cache.port")).getOrElse("80")
  val timeout: Int = Try(properties.getProperty("OTP.query.timeout").toInt).getOrElse(300)
  val ttl: Int = Try(properties.getProperty("OTP.query.ttl").toInt).getOrElse(60)
  val convertTsToMillis: Boolean = Try(properties.getProperty("OTP.query.convertTime").toBoolean).getOrElse(true)
  val maxResultRows: Int = Try(properties.getProperty("OTP.query.maxResultRows").toInt).getOrElse(Int.MaxValue)

  val sid: Int = getCurrentTime

  override def open(): Unit = {}

  override def close(): Unit = {}

  def createJobAndReturnDataset(query: String, connectionInfo: ConnectionInfo, username: String, password: String): Try[Dataset] = {
    Try(Connector(connectionInfo, username, password)) flatMap {
      connector => Try(connector.jobs.create(query, ttl, 0, 0, sid, timeout).getDataset(datasetHost.getOrElse(connectionInfo.host), datasetPort))
    }
  }

  override def interpret(s: String, interpreterContext: InterpreterContext): InterpreterResult = {
    if (host.isDefined && port.isDefined && username.isDefined && password.isDefined) {
      val connectionInfo = ConnectionInfo(host.get, port.get, ssl = false)
      Try(Query(s).setTokens(interpreterContext.getResourcePool)) flatMap {
        query => createJobAndReturnDataset(query.query, connectionInfo, username.get, password.get)
      } flatMap {
        dataset =>
          if (dataset.jsonLines.exists(_.nonEmpty)) Try(parseEvents(dataset.jsonLines.take(maxResultRows), dataset.schema)) else Success("")
      } match {
        case Success(interpretResult) => if (interpretResult.isEmpty) new InterpreterResult(Code.SUCCESS, Type.TEXT, "Resulting dataset is empty")
        else new InterpreterResult(Code.SUCCESS, Type.TABLE, interpretResult)
        case Failure(exception) => new InterpreterResult(Code.ERROR, Type.TEXT, s"""Caught exception '${exception.getMessage}' while interpreting""")
      }
    } else {
      new InterpreterResult(Code.ERROR, Type.TEXT,
        s"""Some interpreter parameters are not specified. Please specify 'OTP.rest.host',
           |'OTP.rest.port', 'OTP.rest.auth.username', 'OTP.rest.auth.password' in interpreter properties.
           |""".stripMargin)
    }
  }

  override def cancel(interpreterContext: InterpreterContext): Unit = {}

  override def getFormType: Interpreter.FormType = FormType.SIMPLE

  override def getProgress(interpreterContext: InterpreterContext): Int = 0

  /**
   * Convert resulting dataset from JSON lines to String in special format for InterpreterResult
   *
   * @param dataset - initial dataset as sequence of JSON Lines
   * @param schema  - schema of dataset
   * @return - dataframe as string; rows are separated with \n, columns - with \t
   */
  def parseEvents(dataset: Seq[String], schema: String): String = {
    val fieldPattern = "`(.+?)`".r
    implicit val formats: DefaultFormats.type = DefaultFormats
    val eventsArrayMap = dataset
      .map(
        x => parse(x)
          .extract[Map[String, Any]]
          .mapValues {
            case vv: List[Any] => vv.toString.replaceAll("List", "")
            case other => other.toString
          }
          .map { case (k, v) => k -> (if ((k == "_time") && convertTsToMillis) convertTimeToMillis(v) else v) })
    val fields = fieldPattern.findAllMatchIn(schema).map(f => f.group(1)).toList
    val headers = fields.mkString("\t")
    val values = eventsArrayMap.map(line => {
      val orderedLine = fields.foldLeft("") {
        (ol, f) =>
          ol + "\t" + line.getOrElse(f, "")
      }
      orderedLine.tail
    }).mkString("\n")
    headers + "\n" + values
  }
}
